packages/flutter_tools/lib/src/build_system/build_system.dart

```diff
@@ -0,0 +1,542 @@
+          .toList(),
+      'stamp': _findStampFile(environment).absolute.path,
+    };
+  }
+
+  /// Locate the stamp file for a particular target name and environment.
+  File _findStampFile(Environment environment) {
+    final String fileName = '$name.stamp';
+    return environment.buildDir.childFile(fileName);
+  }
+
+  static List<SourceFile> _resolveConfiguration(
+      List<Source> config, Environment environment, [bool inputs = true]) {
+    final SourceVisitor collector = SourceVisitor(environment, inputs);
+    for (Source source in config) {
+      source.accept(collector);
+    }
+    return collector.sources;
+  }
+}
+
+
+/// Command line configuration passed to a specific target.
+///
+/// These defines allow arbitrary configuration to be safely used in a build
+/// target.
+///
+/// Example:
+///   flutter assemble foo -d=foo=value -d=foo=value2
+class TargetDefines {
+  TargetDefines(this.defines, this.target);
+
+  /// The rule these defines are for.
+  final String target;
+
+  /// The key value pairs for this specific target.
+  final Map<String, String> defines;
+}
+
+/// The [Environment] defines several constants for use during the build.
+///
+/// The environment contains configuration and file paths that are safe to
+/// depend on and reference during the build.
+///
+/// Example (Good):
+///
+/// Use the environment to determine where to write an output file.
+///
+///    environment.buildDir.childFile('output')
+///      ..createSync()
+///      ..writeAsStringSync('output data');
+///
+/// Example (Bad):
+///
+/// Use a hard-coded path or directory relative to the current working
+/// directory to write an output file.
+///
+///   fs.file('build/linux/out')
+///     ..createSync()
+///     ..writeAsStringSync('output data');
+///
+/// Example (Good):
+///
+/// Using the build mode to produce different output. Note that the invocation
+/// is still responsible for outputting a different file, as defined by the
+/// corresponding output [Source].
+///
+///    if (environment.buildMode == BuildMode.debug) {
+///      environment.buildDir.childFile('debug.output)
+///        ..createSync()
+///        ..writeAsStringSync('debug');
+///    } else {
+///      environment.buildDir.childFile('non_debug.output)
+///        ..createSync()
+///        ..writeAsStringSync('non_debug');
+///    }
+class Environment {
+  /// Create a new [Environment] object.
+  ///
+  /// Only [projectDir] is required. The remaining environment locations have
+  /// defaults based on it.
+  factory Environment({
+    @required Directory projectDir,
+    Directory buildDir,
+    Directory cacheDir,
+    Directory flutterRootDir,
+    Map<String, String> defines = const <String, String>{},
+  }) {
+    assert(projectDir != null);
+    String buildPrefix;
+    if (defines.isNotEmpty) {
+      // Sort the keys by key so that the result is stable.
+      final List<String> keys = defines.keys.toList()..sort();
```

Comment:
Good question.

Changing the build dir would change the location of the filecache and stamps so you would always rebuild anyway.

For the others, supposing that all inputs were adequately captured we shouldn't need to invalidate everything, because if something did change it should be caught by the file hash. The only exception would be changes to the tool itself, which I don't yet capture correctly